src/services/persistent-storage.service.spec.ts

```diff
@@ -0,0 +1,106 @@
+    expect(PersistentStorage).toBeTruthy();
+  });
+
+  it("should instanciate as a singleton", () => {
+    const persistentStorage2 = new PersistentStorage("testing", storage);
+    expect(persistentStorage).toBe(persistentStorage2);
+  });
+
+  it('should init properly', async () => {
+    const initStorage = mock<Storage>({
+      getItem: jest.fn().mockImplementation(() => "testValue"),
+      length: 1,
+      key: jest.fn().mockImplementation(() => "inittestKey"),
+    });
+
+    const initPersistantStorage = new PersistentStorage("init", initStorage);
+
+    const value = await initPersistantStorage.getItem("testKey");
+    expect(value).toBe("testValue");
+    expect(initStorage.getItem).toHaveBeenCalledWith("inittestKey");
+    expect(initPersistantStorage.length).toBe(1);
+
+  });
+
+  it("should set and get a value and sync to localstorage", async () => {
+    await persistentStorage.setItem("test", "test");
+    const setValue = await persistentStorage.getItem("test");
+    expect(setValue).toBe("test");
+  });
+
+  it("should defer setting the value to localstorage until after the JS stack has cleared (performance first)", async () => {
+    await persistentStorage.setItem("test", "test");
+    expect(storage.setItem).toBeCalledTimes(0);
+    await defer(() => {});
+    expect(storage.setItem).toBeCalledTimes(1);
+  });
+
+  it("should not retrieve value from the localstorage", async () => {
+    await persistentStorage.setItem("test", "test");
+    const value = await persistentStorage.getItem("test");
+    expect(value).toBe("test");
+    expect(storage.getItem).toBeCalledTimes(0);
+    await defer(() => {});
+    expect(storage.getItem).toBeCalledTimes(0);
+  });
+
+  it("should get storage size and it should replace values if key is repeated", async () => {
+    await persistentStorage.setItem("test", "test");
+    await persistentStorage.setItem("test2", "test");
+    await persistentStorage.setItem("test3", "test");
+    await persistentStorage.setItem("test", "test");
+    const size = persistentStorage.length;
+    expect(size).toBe(3);
+  });
+
+  it("should clear the storage", async () => {
+    await persistentStorage.setItem("test", "test");
+    await persistentStorage.setItem("test2", "test");
+    await persistentStorage.setItem("test3", "test");
+    await persistentStorage.clear();
+    const size = persistentStorage.length;
+    expect(size).toBe(0);
+  });
+
+  it("should remove item", async () => {
+    await persistentStorage.setItem("test", "test");
+    await persistentStorage.setItem("test2", "test");
+    await persistentStorage.removeItem("test2");
+    expect(await persistentStorage.getItem('test2')).toBe(null);
+    expect(await persistentStorage.getItem('test')).toBe('test');
+  });
+
+  it('should set a prefix on storage', async () => {
+    await persistentStorage.setItem('test', 'test');
+    await defer(() => {});
+    expect(storage.setItem).toBeCalledWith('testingtest', 'test');
```

Comment:
We should look to reference important links between values (this is true for all the tests in this file). In this case, we have 3 values that represent this assertion (prefix, key, and value):

```ts
let prefix: string;
let storage: Storage;
let persistentStorage: PersistentStorage;

beforeAll(() => {
  prefix = "prefix";
  storage = mock<Storage>();
  persistentStorage = new PersistentStorage(prefix, storage);
});

// ...

it('should set a prefix on storage', async () => {
  const key = "key";
  const value = "value";

  await persistentStorage.setItem(key, value);
  await defer(() => {});

  expect(storage.setItem).toBeCalledWith(`${prefix}${key}`, value);
});
```
Having links like this in the code help the reader to understand how things are constructed and make it clear that there are these connections instead of simply a coincidence

Reply:
Sure thing!

Reply:
Another perk to this approach is that if we modify a value in the test, we don't accidently cause them to fail. Like in my example above, if we changed the `prefix`, none of the tests wouldn't break. This avoids potentially suggesting functionality has regressed